### PR TITLE
route relay via tor proxy

### LINKED ISSUE #696: Encrypt communication to nostr relay
Currently posting an offer to a nostr relay exposes our IP address due to the communication not being proxied. This is a source of privacy leak for the makers.

### PR DESCRIPTION
**Pull request by 0xEgao**
# Pull Request

## Related Issue(s)
Fixes #696

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor / performance improvement
- [ ] Documentation update only
- [ ] CI / Docker / Build changes
- [ ] Other (please describe):

## Protocol Version(s) Affected
- [ ] Legacy (ECDSA — `messages.rs`, `contract.rs`, `handlers.rs`)
- [ ] Taproot-Musig2 (`messages2.rs`, `contract2.rs`, `handlers2.rs`)
- [ ] Both
- [x] Neither (infrastructure / tooling only)

> **Note:** When modifying protocol logic, changes must be made to **both** versions unless the
> scope is explicitly version-specific.

## Affected Component(s)
- [ ] **makerd** (background daemon)
- [ ] **taker** (CLI client)
- [ ] **maker-cli** (command-line tool)
- [ ] Core protocol / cryptography / library
- [x] **watch_tower** (contract monitoring & recovery)
- [ ] Docker / deployment scripts
- [ ] Tests / test framework
- [ ] Documentation (`docs/`)
- [ ] Other (please specify):

### Code Quality
- [x] I ran `cargo +nightly fmt --all` and committed the result
- [x] I ran `cargo +stable clippy --all-features --lib --bins --tests -- -D warnings` with zero warnings
- [x] I ran `cargo +stable clippy --examples -- -D warnings` with zero warnings
- [x] I ran `RUSTDOCFLAGS="-D warnings" cargo +nightly doc --all-features --document-private-items --no-deps` with zero warnings
- [x] Pre-commit git hook passes (`ln -s ../../git_hooks/pre-commit .git/hooks/pre-commit` if not already set)

### Testing
- [x] All unit tests pass (`cargo test`)
- [x] Integration tests pass (`cargo test --features integration-test`)
- [x] Changes were manually tested on **signet**
- [x] End-to-end maker ↔ taker swap tested (where applicable)
- [x] Test-only code is gated behind `#[cfg(feature = "integration-test")]`

### Documentation
- [ ] Relevant files in the `docs/` folder were updated
- [ ] Complex logic is commented

### Security & Privacy (Critical)
- [ ] This change preserves **trustlessness** and **atomic swap guarantees**
- [ ] No regression in **sybil resistance** or **fidelity bond** logic
- [ ] Tor anonymity and P2P message flow were reviewed
- [ ] No new attack vectors or trust assumptions introduced
- [ ] Edge cases and error handling considered
- [ ] ZMQ subscription integrity (raw tx/block feed) is unaffected
- [ ] `integration-test` feature flag is not reachable in production code paths

## How to Test

```bash
# Standard swap (good baseline)
cargo test --test standard_swap --features integration-test -- --nocapture

# Abort scenarios
cargo test --test abort1 --features integration-test -- --nocapture
cargo test --test taproot_taker_abort1 --features integration-test -- --nocapture

# Malicious behavior & recovery
cargo test --test malice1 --features integration-test -- --nocapture
cargo test --test taproot_timelock_recovery --features integration-test -- --nocapture

# Or run the full suite
cargo test --features integration-test
```

## Summary by CodeRabbit

* **New Features**
  * Tor/SOCKS5 proxy support with optional authentication for Nostr relay connections.
  * More robust WebSocket handling: explicit TLS validation, TCP read/write timeouts, and relay connection retries.
  * Nostr discovery and relay sessions can be routed through Tor where configured.

* **Bug Fixes**
  * Startup validation ensures wallet/network settings and runtime config are consistent.

### CODE REVIEW

File: src/maker/server.rs [hunk 1]
```diff
@@ -229,14 +229,23 @@ fn spawn_nostr_broadcast_thread(
 
     let maker_clone = Arc::clone(maker);
     let relays = maker.nostr_relays.clone();
+    let nostr_network = maker
+        .wallet
+        .read()
+        .map(|w| w.store.network)
+        .unwrap_or_else(|_| maker.config.network);
```
Comment (0xEgao) [inline src/maker/server.rs]: #824 introduced different network kinds, but there was no network check for maker before broadcasting bond, this was leading to failure in offer book syncing on network other than `regtest`. Fixed that now.

File: src/watch_tower/watcher.rs [hunk 2]
```diff
@@ -139,22 +142,29 @@ impl<R: Role> Watcher<R> {
...
+                            discovery_shutdown.clone(),
+                            initial_sync_complete,
+                            &nostr_relays,
+                            nostr_tor_config,
+                        ) {
+                            log::error!("Discovery thread failed: {:?}", e);
+                        }
+                    });
+                } else {
+                    log::warn!("Nostr discovery skipped: Tor config not provided");
```
Comment (mojoX911) [inline src/watch_tower/watcher.rs]: Do we need this if cfg!() filters are already provided? 

In prod we will always have tor configs. In tests, discovery should still run without Tor. So where does this clause apply?

File: src/maker/server.rs [hunk 3]
```diff
@@ -256,9 +265,13 @@ fn spawn_nostr_broadcast_thread(
...
 
-                if let Err(e) =
-                    broadcast_bond_on_nostr(fidelity.clone(), maker_clone.config.network, &relays)
-                {
+                if let Err(e) = broadcast_bond_on_nostr(
+                    fidelity.clone(),
+                    nostr_network,
+                    &relays,
+                    maker_clone.config.socks_port,
+                    &maker_clone.config.tor_auth_password,
```
Comment (mojoX911) [inline src/maker/server.rs]: nostr_network, socks port, tor password they all come form the config, So why not just pass on the &config and derive them inline? function sig will be simpler, and derivation logic will be localized.

**Review by mojoX911 — CHANGES_REQUESTED** [review]

File: src/nostr_coinswap.rs [hunk 4]
```diff
@@ -39,11 +44,60 @@ pub fn coinswap_kind(network: Network) -> u16 {
...
+
+    tcp.set_read_timeout(Some(Duration::from_secs(30)))?;
+    tcp.set_write_timeout(Some(Duration::from_secs(30)))?;
+    match tungstenite::client_tls_with_config(relay_url, tcp, None, None) {
+        Ok((ws, _)) => Ok(ws),
+        Err(tungstenite::HandshakeError::Failure(e)) => Err(e),
+        Err(tungstenite::HandshakeError::Interrupted(_)) => {
+            Err(io::Error::other("tls handshake interrupted").into())
+        }
+    }
```
[the remark this answers is not in the document]
Comment (0xEgao) [reply src/nostr_coinswap.rs:94]: Not an issue, `client_tls_with_config` returns `MaybeTlsStream` which can handle both plain(ws) and tls(wss).

File: src/watch_tower/watcher.rs (hunk 2 above)
Comment (0xEgao) [reply src/watch_tower/watcher.rs]: Yeah it is redundant, removed it.

File: src/maker/server.rs (hunk 3 above)
Comment (0xEgao) [reply src/maker/server.rs]: Done, as well added the network check logic to maker server initialisation, to avoid `MakerServerConfig` always having `regtest` as network in it.

File: src/nostr_coinswap.rs [hunk 5]
```diff
@@ -39,15 +47,62 @@ pub fn coinswap_kind(network: Network) -> u16 {
...
+    } else {
+        Socks5Stream::connect_with_password(
+            socks_addr.as_str(),
+            target_addr.as_str(),
+            "",
+            tor_auth_password,
+        )
+    }
+    .map_err(|e| io::Error::other(e.to_string()))?
+    .into_inner();
```
[the remark this answers is not in the document]
Comment (0xEgao) [reply src/nostr_coinswap.rs:84]: kept host for now

**Review by mojoX911 — APPROVED** [review]
Ack
